standards/SPIFFE_Federation.md

```diff
@@ -0,0 +1,316 @@
+
+## 4. SPIFFE Bundle Endpoint
+A SPIFFE bundle endpoint is a resource (represented by a URL) that serves a copy of a SPIFFE bundle for a trust domain. SPIFFE control planes may both expose and consume these endpoints in order to transfer bundles between themselves, thereby achieving federation.
+
+The semantics of the SPIFFE bundle endpoint are similar to the `jwks_uri` mechanism defined in the OpenID Connect specification in that the bundle endpoint contains one or more public cryptographic keys used by a trust domain (for certifying identities within the trust domain). The bundle endpoint is an HTTPS URL that responds to an HTTP GET with a SPIFFE bundle. For more information about SPIFFE bundles and how they are encoded, please see the [SPIFFE Trust Domain and Bundle][3] specification.
+
+### 4.1. Adding and Removing Keys
+Operators of a trust domain MAY introduce or remove keys used to issue SVIDs within the trust domain as needed (e.g. as part of an internal key rotation process). When adding new keys, an updated trust bundle containing the keys SHOULD be published at the bundle endpoint sufficiently in advance that foreign trust domains have an opportunity to retrieve and internally disseminate the new bundle contents; the recommended advance time is 3-5 times the bundle’s `spiffe_refresh_hint`. At a minimum, new keys MUST be published at the bundle endpoint prior to the keys being used to issue SVIDs. 
+
+Deprecated keys SHOULD be removed from the trust bundle after a trust domain no longer has any active, valid SVIDs issued from those keys. Failure to follow these recommendations as keys are added to and removed from the bundle could result in transient cross-domain authentication failures. Requirements for updating the trust bundle do not apply to keys used to issue SVIDs for internal use only.
+
+Clients SHOULD periodically poll the endpoint for updates because the contents are expected to [change over time][4] - key validity periods on the order of weeks or even days is commonplace. Clients SHOULD poll at a frequency equal to the value of the bundle’s `spiffe_refresh_hint`, in seconds. If not set, a reasonably low default value should apply - five minutes is recommended.
+
+### 4.2. Managing Fetched Bundles
+Clients of the bundle endpoint MUST store the latest SPIFFE bundle each time it is retrieved. The sequence number field of the trust bundle SHOULD be used when comparing the freshness or ordering of two trust bundles. If the trust bundle omits the sequence number, operators SHOULD consider the most recently retrieved bundle to be up-to-date.
```

Comment:
MUST or SHOULD ?

What breaks if a client doesn't store the SPIFFE bundle?

Reply:
To clarify, the suggestion here is to replace "MUST store the latest" with "SHOULD store the latest".

Reply:
I'm ok with SHOULD.

If the client doesn't store the bundle at all, then it will need to issue another GET each time it is needed.  There are probably some legit use cases for that, like a workload that needs to validate an SVID once a day.

If the client stores the bundle, but doesn't store the latest bundle, then it could fall behind and fail validation of SVIDs.

Reply:
+1 for SHOULD

Spike's example of validating an SVID once a day is a good one (cron jobs come to mind). I think this could also be important for safety. For example, if a client notices that the latest SPIFFE bundle has no keys present, or some similarly suspect change, they may choose to ignore it.